src/transports/ipc.rs

```diff
@@ -156,44 +157,91 @@ impl Ipc {
 }
 
 impl Transport for Ipc {
-  type Out = IpcTask;
+  type Out = IpcTask<fn (Vec<RpcResult<rpc::Value>>) -> RpcResult<rpc::Value>>;
 
-  fn execute(&self, method: &str, params: Vec<RpcValue>) -> Self::Out {
+  fn prepare(&self, method: &str, params: Vec<rpc::Value>) -> (RequestId, rpc::Call) {
     let id = self.id.fetch_add(1, atomic::Ordering::AcqRel);
     let request = helpers::build_request(id, method, params);
+
+    (id, request)
+  }
+
+  fn send(&self, id: RequestId, request: rpc::Call) -> Self::Out {
+    let request = helpers::to_string(&rpc::Request::Single(request));
     debug!("Calling: {}", request);
 
-    // When the response is ready
     let (tx, rx) = futures::oneshot();
     self.pending.lock().insert(id, tx);
-    let result = self.write_sender.lock().send(request.into_bytes());
+    let result = {
+      let mut sender = self.write_sender.lock();
+      (*sender).send(request.into_bytes())
+    };
+
+    IpcTask {
+      state: IpcTaskState::Sending(Some(result), rx),
+      extract: single_response as fn(Vec<RpcResult<rpc::Value>>) -> RpcResult<rpc::Value>,
+    }
+  }
+}
+
+fn single_response(response: Vec<RpcResult<rpc::Value>>) -> RpcResult<rpc::Value> {
+  match response.into_iter().next() {
+    Some(res) => res,
+    None => Err(RpcError::Transport(format!("Expected single response got empty batch."))),
+  }
+}
+
+fn batch_response(response: Vec<RpcResult<rpc::Value>>) -> RpcResult<Vec<RpcResult<rpc::Value>>> {
+  Ok(response)
+}
+
+impl BatchTransport for Ipc {
+  type Batch = IpcTask<fn(Vec<RpcResult<rpc::Value>>) -> RpcResult<Vec<RpcResult<rpc::Value>>>>;
+
+  fn send_batch(&self, requests: Vec<(RequestId, rpc::Call)>) -> Self::Batch {
+    let id = requests.get(0).map(|x| x.0).unwrap_or(0);
+    let requests = requests.into_iter().map(|x| x.1).collect();
+    let request = helpers::to_string(&rpc::Request::Batch(requests));
+    debug!("Calling: {}", request);
+
+    let (tx, rx) = futures::oneshot();
+    self.pending.lock().insert(id, tx);
+    let result = {
+      let mut sender = self.write_sender.lock();
+      (*sender).send(request.into_bytes())
+    };
 
     IpcTask {
       state: IpcTaskState::Sending(Some(result), rx),
+      extract: batch_response as fn(Vec<RpcResult<rpc::Value>>) -> RpcResult<Vec<RpcResult<rpc::Value>>>,
```

Comment:
you don't technically need a separate `batch_response` function for this: enum variants are their own functions so I think `Ok as fn(...) -> ...` would work